src/WallpaperEngine/Render/CWallpaperState.cpp

```diff
@@ -0,0 +1,170 @@
+#include "CWallpaperState.h"
+#include <algorithm>
+#include "WallpaperEngine/Logging/CLog.h"
+
+using namespace WallpaperEngine::Render;
+
+// Reset UVs to 0/1 values
+void CWallpaperState::resetUVs(){
+    this->UVs.ustart=0;
+    this->UVs.uend=1;
+    if (vflip){
+        this->UVs.vstart = 0.0f;
+        this->UVs.vend = 1.0f;
+    }
+    else{
+        this->UVs.vstart=1.0f;
+        this->UVs.vend=0.0f;
+    }
+}
+
+// Update Us coordinates for current viewport and projection
+void CWallpaperState::updateUs(const int& projectionWidth, const int& projectionHeight){
+    const float viewportWidth  = this->getViewportWidth();
+    const float viewportHeight = this->getViewportHeight();
+    int newWidth = viewportHeight  / projectionHeight * projectionWidth;
+    float newCenter = newWidth / 2.0f;
+    float viewportCenter = viewportWidth / 2.0;
+
+    float left = newCenter - viewportCenter;
+    float right = newCenter + viewportCenter;
+
+    this->UVs.ustart = left / newWidth;
+    this->UVs.uend   = right / newWidth;
+}
+
+// Update Vs coordinates for current viewport and projection 
+void CWallpaperState::updateVs(const int& projectionWidth, const int& projectionHeight){
+    const float viewportWidth  = this->getViewportWidth();
+    const float viewportHeight =  this->getViewportHeight();
+    int newHeight = viewportWidth / projectionWidth * projectionHeight;
+    float newCenter = newHeight / 2.0f;
+    float viewportCenter = viewportHeight / 2.0;
+
+    float down = newCenter - viewportCenter;
+    float up = newCenter + viewportCenter;
+
+    if (vflip)
+    {
+        this->UVs.vstart = down / newHeight;
+        this->UVs.vend = up / newHeight;
+    }
+    else
+    {
+        this->UVs.vstart = up / newHeight;
+        this->UVs.vend = down / newHeight;
+    }
+}
+
+
+template<> void CWallpaperState::updateTextureUVs<CWallpaperState::TextureUVsScaling::StretchUVs>(){
+    this->resetUVs();
+}
+
+template<> void CWallpaperState::updateTextureUVs<CWallpaperState::TextureUVsScaling::ZoomFillUVs>(){
+    this->resetUVs();
+
+    const int viewportWidth  = this->getViewportWidth();
+    const int viewportHeight =  this->getViewportHeight();
+    uint32_t projectionWidth = this->getProjectionWidth();
+    uint32_t projectionHeight = this->getProjectionHeight();
+
+    const float m1 = static_cast<float>(viewportWidth) / projectionWidth;
+    const float m2 = static_cast<float>(viewportHeight) / projectionHeight;
+    const float m = std::max(m1,m2);
```

Comment:
This (line 74) is the only difference with  `ZoomFitUVs`(line 98) specialization